dali/operators/generic/histogram/histogram.cc

```diff
@@ -0,0 +1,685 @@
+#include "dali/kernels/transpose/transpose.h"
+#include "dali/pipeline/operator/op_schema.h"
+
+
+#include "dali/core/tensor_shape_print.h"
+
+#include "opencv2/imgproc.hpp"
+#include "opencv2/core/core_c.h"
+
+#include <iostream>
+
+using namespace dali;
+using namespace dali::hist_detail;
+
+#define id_(x) x
+
+#define HistogramOpName id_(histogram__Histogram)
+#define UniformHistogramOpName id_(histogram__UniformHistogram)
+
+#define str_next_(x) #x
+#define str_(x) str_next_(x)
+
+namespace {
+
+constexpr const char histogramOpString[] = str_(HistogramOpName);
+constexpr const char unifromHistogramOpString[] = str_(UniformHistogramOpName);
+
+static constexpr int HISTOGRAM_MAX_CH = CV_MAX_DIM;
+
+std::vector<float> GetFlattenedRanges(int sample, const workspace_t<CPUBackend> &ws) {
+  std::vector<float> ranges;
+
+  for (int r = 1; r < ws.NumInput(); ++r) {
+    auto &dim_ranges = ws.template Input<CPUBackend>(r);
+    auto range_view = view<const float>(dim_ranges);
+    for (int i = 0; i < range_view[sample].num_elements(); ++i) {
+      ranges.push_back(range_view.tensor_data(sample)[i]);
+    }
+  }
+  return ranges;
+}
+
+template <typename Ty_>
+struct CVMatType {
+  static int get(int) {
+    DALI_ENFORCE(false, "Unreachable - invalid type");
+  }
+};
+
+template <>
+struct CVMatType<uint8_t> {
+  static int get(int nchannel) noexcept {
+    return CV_MAKETYPE(CV_8U, nchannel);
+  }
+};
+
+template <>
+struct CVMatType<uint16_t> {
+  static int get(int nchannel) noexcept {
+    return CV_MAKETYPE(CV_16U, nchannel);
+  }
+};
+
+template <>
+struct CVMatType<float> {
+  static int get(int nchannel) noexcept {
+    return CV_MAKETYPE(CV_32F, nchannel);
+  }
+};
+
+template <>
+struct CVMatType<double> {
+  static int get(int nchannel) noexcept {
+    return CV_MAKETYPE(CV_64F, nchannel);
+  }
+};
+
+template <typename Type, typename ScratchAlloc, typename Coll>
+TensorListView<StorageCPU, const Type> transpose_view(
+    dali::ThreadPool &thread_pool, ScratchAlloc &scratch,
+    const TensorListView<StorageCPU, const Type> &in_view, const Coll &transpose_axes_order) {
+  const auto &in_shapes = in_view.shape;
+
+  TensorListShape<> transposed_shapes;
+  permute_dims(transposed_shapes, in_shapes, transpose_axes_order);
+  std::vector<Type *> tmp_pointers;
+  tmp_pointers.reserve(transposed_shapes.num_samples());
+
+  for (int i = 0; i < transposed_shapes.num_samples(); ++i) {
+    auto tmp = scratch.template AllocTensor<mm::memory_kind::host, Type>(transposed_shapes[i]);
+    tmp_pointers.push_back(tmp.data);
+  }
+
+  TensorListView<StorageCPU, Type> transpose_out_view(std::move(tmp_pointers),
+                                                      std::move(transposed_shapes));
+
+  for (int i = 0; i < transpose_out_view.num_samples(); ++i) {
+    thread_pool.AddWork([&, i](int thread_id) {
+      auto perm = make_span(transpose_axes_order);
+      kernels::Transpose(transpose_out_view[i], in_view[i], perm);
```

Comment:
maybe `kernels::TransposeGrouped` as it can automatically simplify the shapes when applicable?